pkg/api/jobartifacts/manager.go

```diff
@@ -0,0 +1,250 @@
+	ctx, cancelTimeout := context.WithTimeout(ctx, jobRunQueryTimeout)
+	defer cancelTimeout()
+
+	// send all the requests with a return channel
+	jobRunResponseChan := make(chan jobRunResponse)
+	for _, id := range query.JobRunIDs {
+		m.jobRunChan <- jobRunRequest{
+			jobRun:      id,
+			query:       query,
+			jobRunsChan: jobRunResponseChan,
+			ctx:         ctx,
+		}
+	}
+
+	// wait for the responses
+	remaining := sets.NewInt64(query.JobRunIDs...)
+responseLoop:
+	for {
+		select {
+		case <-ctx.Done(): // timed out
+			break responseLoop
+		case res := <-jobRunResponseChan:
+			remaining.Delete(res.jobRun) // we have seen it, no longer remaining
+			if res.error != nil {
+				result.Errors = append(result.Errors, JobRunError{
+					ID:    strconv.FormatInt(res.jobRun, 10),
+					Error: res.error.Error(),
+				})
+			} else {
+				result.JobRuns = append(result.JobRuns, res.response)
+			}
+			if len(remaining) == 0 {
+				break responseLoop // we have received a response for every request
+			}
+		}
+	}
+
+	// fill in errors for any that went missing
+	for it := range remaining {
+		result.Errors = append(result.Errors, JobRunError{
+			ID:    strconv.FormatInt(it, 10),
+			Error: fmt.Sprintf("request did not complete within %s", jobRunQueryTimeout),
+		})
+	}
+
+	// sort things for consistency
+	slices.SortFunc(result.JobRuns, func(a, b JobRun) int { // sort runs by ID ascending
+		return strings.Compare(a.ID, b.ID)
+	})
+	slices.SortFunc(result.Errors, func(a, b JobRunError) int { // sort errors by ID ascending
+		return strings.Compare(a.ID, b.ID)
+	})
+	return
+}
+
+// a jobRunRequest channels a single job run to be processed and returned with matching file paths
+type jobRunRequest struct {
+	jobRun      int64
+	query       *JobArtifactQuery
+	jobRunsChan chan jobRunResponse // when processing is done, send response to this channel
+	ctx         context.Context     // enable cancellation/timeout of query
+}
+
+// jobRunReponse channels the response for a single job run that was processed
+type jobRunResponse struct {
+	jobRun   int64
+	response JobRun
+	error    error
+}
+
+func (m *Manager) jobRunWorker(ctx context.Context) {
+	defer m.jobRunWorkers.Done()
+	logger := log.WithField("func", "Manager.jobRunWorker")
+	for {
+		select {
+		case <-ctx.Done(): // shutting down the workers
+			logger.Debugf("Shutting down per context: %v", ctx.Err())
+			return
+		case req := <-m.jobRunChan:
+			jobLog := logger.WithField("jobRunId", req.jobRun).WithContext(req.ctx)
+			jobLog.Debug("Received from jobRunChan")
+			select {
+			case <-req.ctx.Done(): // query is starting too late, request already cancelled
+				jobLog.WithError(ctx.Err()).Warn("Context aborted request for job run artifacts - getting backlogged?")
+				req.jobRunsChan <- jobRunResponse{
+					jobRun: req.jobRun,
+					error:  fmt.Errorf("job run request was not serviced soon enough: %v", req.ctx.Err()),
+				}
```

Comment:
Is a continue missing here?

Reply:
I don't think so? Once the case is done, should just finish the select and drop out to the next `for` iteration...

Anyway I tried adding it explicitly and no difference.